kernel/pci/src/lib.rs

```diff
@@ -221,74 +263,98 @@ impl PciLocation {
     /// The least two significant bits of offset are masked, so it's 4-byte aligned addressing,
     /// which makes sense since we read PCI data (from the configuration space) in 32-bit chunks.
     fn pci_address(self, offset: u8) -> u32 {
-        ((self.bus  as u32) << 16) | 
-        ((self.slot as u32) << 11) | 
-        ((self.func as u32) <<  8) | 
-        ((offset as u32) & (PCI_CONFIG_ADDRESS_OFFSET_MASK as u32)) | 
-        0x8000_0000
+        BASE_OFFSET |
+        ((self.bus  as u32) << 16) |
+        ((self.slot as u32) << 11) |
+        ((self.func as u32) <<  8) |
+        (offset as u32)
     }
 
-    /// read 32-bit data at the specified `offset` from the PCI device specified by the given `bus`, `slot`, `func` set.
-    fn pci_read_32(&self, offset: u8) -> u32 {
-        unsafe { 
-            PCI_CONFIG_ADDRESS_PORT.lock().write(self.pci_address(offset)); 
+    fn pci_read_32_addr(&self, address: u32) -> u32 {
+        #[cfg(target_arch = "x86_64")] {
+            let shift = (address & (!PCI_CONFIG_ADDRESS_OFFSET_MASK)) * 8;
+            let aligned_address = address & PCI_CONFIG_ADDRESS_OFFSET_MASK;
+            unsafe { 
+                PCI_CONFIG_ADDRESS_PORT.lock().write(aligned_address);
+            }
+            PCI_CONFIG_DATA_PORT.lock().read() >> shift
+        }
+
+        #[cfg(target_arch = "aarch64")] {
+            let config_space = PCI_CONFIG_SPACE.lock();
+            let config_space = config_space.deref().as_ref().unwrap();
+            let address = address as usize;
+            let slice = &config_space[address..][..size_of::<u32>()];
+            let mut value = [0; 4];
+            value.copy_from_slice(slice);
+            u32::from_ne_bytes(value)
         }
-        Self::read_data_port() >> ((offset & (!PCI_CONFIG_ADDRESS_OFFSET_MASK)) * 8)
+    }
+
+    /// Read 32-bit data at the specified `offset` from this PCI device.
+    fn pci_read_32(&self, offset: u8) -> u32 {
+        self.pci_read_32_addr(self.pci_address(offset))
     }
 
     /// Read 16-bit data at the specified `offset` from this PCI device.
     fn pci_read_16(&self, offset: u8) -> u16 {
         self.pci_read_32(offset) as u16
     } 
 
-    /// Read 8-bit data at the specified `offset` from the PCI device.
+    /// Read 8-bit data at the specified `offset` from this PCI device.
     fn pci_read_8(&self, offset: u8) -> u8 {
         self.pci_read_32(offset) as u8
     }
 
-    /// Write 32-bit data to the specified `offset` for the PCI device.
-    fn pci_write(&self, offset: u8, value: u32) {
-        unsafe {
-            PCI_CONFIG_ADDRESS_PORT.lock().write(self.pci_address(offset)); 
-            Self::write_data_port((value) << ((offset & 2) * 8));
+    fn pci_write_addr(&self, address: u32, value: u32) {
```

Comment:
why do we need a new function for this? Perhaps I don't understand what this is for, but what's the difference between an "address" (`u32`) and and "offset" (`u8`)? 

This function seems like it overlaps with `pci_write_32`

Reply:
Yep, I intended this to allow successive calls to pci_read/write to keep the translated address in their stack frame between calls, but it's probably already happening on optimized builds. I'll restore the original API.